### PR TITLE
wasm-bindgen-cli: init at 0.2.51

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

CLI for `wasm-bindgen`, a tool that helps to interact between WASM
modules written in Rust and JS-Code. This contains only the CLI[1], the
libraries are usually included via `Cargo.toml` in a project.

[1] https://crates.io/crates/wasm-bindgen-cli

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/development/tools/wasm-bindgen-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  };
+
+  buildInputs = [ openssl ];
+  nativeBuildInputs = [ pkgconfig ];
+
+  cargoSha256 = "0a7w99az4yqlzi6nh0hyrpalvps6nzc5cbzgrimf1vg7kh16plxr";
+  cargoPatches = [ ./0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch ];
+  cargoBuildFlags = [ "-p" pname ];
+
+  meta = with lib; {
```
Comment (mmahut) [inline pkgs/development/tools/wasm-bindgen-cli/default.nix]: Please add platforms.

File: pkgs/development/tools/wasm-bindgen-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{ rustPlatform, fetchFromGitHub, lib, openssl, pkgconfig, ... }:
```
Comment (mmahut) [inline pkgs/development/tools/wasm-bindgen-cli/default.nix]:
```suggestion
{ rustPlatform, fetchFromGitHub, lib, openssl, pkgconfig, darwin }:
```

File: pkgs/development/tools/wasm-bindgen-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  version = "0.2.50";
+
+  src = fetchFromGitHub {
+    owner = "rustwasm";
+    repo = "wasm-bindgen";
+    rev = version;
+    sha256 = "1axlp4qa2fi79giqwcq105qb9n0znfmknrmnqxy6vdzqzgfszqfq";
+  };
+
+  buildInputs = [ openssl ];
```
Comment (mmahut) [inline pkgs/development/tools/wasm-bindgen-cli/default.nix]: Fails to build on darwin:

```
"Security" "-lSystem" "-lresolv" "-lc" "-lm"
  = note: ld: framework not found Security
          clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
error: aborting due to previous error
```

```suggestion
  buildInputs = [ openssl ] ++ stdenv.lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.Security ];
```

**Comment by Ma27** [comment]
@mmahut fixed :)

**Comment by disassembler** [comment]
@GrahamcOfBorg build wasm-bindgen-cli

**Comment by Ma27** [comment]
Disabled darwin build for now as I have no time and interest in maintaining a stable darwin build for this package as well. In case somebody is interested in darwin suport, feel free to file a patch :)

**Comment by worldofpeace** [comment]
> Disabled darwin build for now as I have no time and interest in maintaining a stable darwin build for this package as well. In case somebody is interested in darwin suport, feel free to file a patch :)

That's fine :+1: As the package maintainer it's best to support what you actually can support.

File: pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch [hunk 4]
```diff
@@ -0,0 +1,2254 @@
+From ab4953af561946fbbe3a7d6ce0c41931dcb28d4e Mon Sep 17 00:00:00 2001
+From: Maximilian Bosch <maximilian@mbosch.me>
+Date: Thu, 29 Aug 2019 21:23:14 +0200
+Subject: [PATCH] Add cargo.lock for `rustPlatform.buildRustPackage` in nixpkgs
```
Comment (worldofpeace) [inline pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch]: Does upstream not commit the lock file?

Comment (Ma27) [reply pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch]: No (at least I didn't find a suitable one in the repo we fetch in this expression). Or am I misunderstanding you? :)

Comment (worldofpeace) [reply pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch]: Understood exactly. This won't block merge, but can you submit a PR to them commiting the lock file? Kinda weird they don't have one in their github repo.

**Comment by B4dM4n** [comment]
I made a patch to support macOS builds, if you want to merge it:

https://github.com/Ma27/nixpkgs/compare/package-wasm-bindgen-cli...B4dM4n:wasm-bindgen-cli-darwin

Also note, when building the package I got a `cargoSha256` mismatch. Not sure what changed, but the old `cargoSha256 = "0a7w99az4yqlzi6nh0hyrpalvps6nzc5cbzgrimf1vg7kh16plxr"` is not working anymore.

**Comment by B4dM4n** [comment]
I dug a bit deeper and the issue probably arises from the fact that `wasm-bindgen-cli-0.2.50-vendor` is a fixed-output derivation. 
This implies, that changing the patch that modifies the `Cargo.lock` does not trigger a rebuild of `wasm-bindgen-cli-0.2.50-vendor`, if a old version with a matching hash is in the store or available from substituters.

So whenever the inputs of `fetchcargo` change (currently `src srcs sourceRoot cargoUpdateHook cargoPatches`), the `cargoSha256` needs to be invalidated.

**Comment by Ma27** [comment]
Yeah, this is one of the downsides of this approach with fixed-output derivations. I'm currently investigating a different issue with wasm-bindgen-cli 0.2.50 which is why this is a bit delayed now... Will report back as soon as I have something usable :)

File: pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch (hunk 4 above)
Comment (Ma27) [reply pkgs/development/tools/wasm-bindgen-cli/0001-Add-cargo.lock-for-rustPlatform.buildRustPackage-in-.patch]: Reported in https://github.com/rustwasm/wasm-bindgen/issues/1819.

**Comment by Ma27** [comment]
Now having 0.2.51.

@B4dM4n thanks for the fix! Squashed it together and added you as co-author :)

@ofborg build wasm-bindgen-cli

**Comment by Ma27** [comment]
Hmm, it actually seems as the cargo hash isn't deterministic (I actually rechecked it locally). Let's see what the x86_64-linux and darwin build does. Will take a look into this.

**Comment by Ma27** [comment]
Aah.... it seems as some cargo-vendor changes were made, this branch is rather old. After rebasing onto latest master I get the same hash now. Currently rebuilding, will push a fix then :)

**Comment by Ma27** [comment]
@ofborg build wasm-bindgen-cli

**Pull request merged**